data/kernels/extended.cl

```diff
@@ -877,7 +877,7 @@ colorbalancergb (read_only image2d_t in, write_only image2d_t out,
   // factorization of : (RGB[c] * (1.f - alpha) + RGB[c] * d->shadows[c] * alpha) * (1.f - beta)  + RGB[c] * d->highlights[c] * beta;
 
   // midtones : power with sign preservation
-  RGB = sign(RGB) * native_powr(fabs(RGB) / white_fulcrum, midtones) * white_fulcrum;
+  RGB = fmax(0.0f, sign(RGB) * native_powr(fabs(RGB) / white_fulcrum, midtones) * white_fulcrum);
```

Comment:
And indeed, this seems equivalent to removing `sign(RGB)` as we take fabs(RGB) anyway.

Reply:
Tested and the following patch also fixes the issue:

```
diff --git a/data/kernels/extended.cl b/data/kernels/extended.cl
index 7ce955268b..273d274d91 100644
--- a/data/kernels/extended.cl
+++ b/data/kernels/extended.cl
@@ -877,7 +877,7 @@ colorbalancergb (read_only image2d_t in, write_only image2d_t out,
   // factorization of : (RGB[c] * (1.f - alpha) + RGB[c] * d->shadows[c] * alpha) * (1.f - beta)  + RGB[c] * d->highlights[c] * beta;
 
   // midtones : power with sign preservation
-  RGB = sign(RGB) * native_powr(fabs(RGB) / white_fulcrum, midtones) * white_fulcrum;
+  RGB = native_powr(fabs(RGB) / white_fulcrum, midtones) * white_fulcrum;
 
   // for the non-linear ops we need to go in Yrg again because RGB doesn't preserve color
   LMS = gradingRGB_to_LMS(RGB);

```

But is that correct? I mean the code and comment about preserving the sign really mean that this was wanter :)

Reply:
Let me investigate source in more depth. We do some sort of sign preserving gamma correction here so very neg values in RGB result in neg values too.

Reply:
1. I now understand why there were no artefacts in the JzAzBz code, there the conversion via `dt_XYZ_2_JzAzBz` takes care
2. we can & should do the "input-sanity check" later in the specific code for UCS before `dt_XYZ_to_xyY`
3. Surprise,  surprise - spotted a possible div by zero in this code too. Also in OpenCL explaining why some drivers don't like this module. (BTW i checked afterwards, we still need the input sanity check) Also checked the JzAzBz code does not have this problem.